### PR TITLE
[Add] frequency_manager to check log "readiness"

### PR DESCRIPTION
**Pull request by rahul-tuli**
This PR adds a `FrequencyManager` class

Responsibilities of this class:
 - Ascertain if we should log at the current step or not

Based off of #1924

Testing: Added automated tests

### CODE REVIEW

File: src/sparseml/core/logger/utils/frequency_manager.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import Union
+
+
+__all__ = ["FrequencyManager"]
+
+LogStepType = Union[int, float, None]
```
Comment (bfineran) [inline src/sparseml/core/logger/utils/frequency_manager.py:21]: Why do we call out int? Is there any int specific logic?

**Review by bfineran — CHANGES_REQUESTED** [review]
Shouldn't the frequency be set differently based on additional args? Ie we only care about model update if we're in on_change mode

File: src/sparseml/core/logger/utils/frequency_manager.py (hunk 1 above)
Comment (rahul-tuli) [reply src/sparseml/core/logger/utils/frequency_manager.py:21]: Int will be needed for `optimizer_step` and `batch` frequency types; the support for that will come through in a different PR

**Comment by rahul-tuli** [comment]
> Shouldn't the frequency be set differently based on additional args? Ie we only care about model update if we're in on_change mode

Support for different modes, like `on_change` and `exact` will come through in a different PR.

**Review by bfineran — APPROVED** [review]

**Review by dbogunowicz — APPROVED** [review]

**Pull request merged**